File: crates/piper-bundle/stdlib/_ast.py
PyCF_ONLY_AST = 1024
PyCF_TYPE_COMMENTS = 4096
PyCF_ALLOW_TOP_LEVEL_AWAIT = 8192
PyCF_OPTIMIZED_AST = 32768


class AST:
    _fields = ()
    _attributes = ()
    _field_types = {}

    def __init__(self, *args, **kwargs):
        if len(args) > len(self._fields):
            raise TypeError(f"{type(self).__name__} constructor takes at most {len(self._fields)} positional arguments")
        for name, value in zip(self._fields, args):
            setattr(self, name, value)
        for name, value in kwargs.items():
            setattr(self, name, value)


def _make(name, base, fields=(), attributes=None):
    if attributes is None:
        attributes = base._attributes
    return type(name, (base,), {
        "_fields": fields,
        "_attributes": attributes,
        "_field_types": {},
        "__module__": "ast",
    })


mod = _make("mod", AST)
stmt = _make("stmt", AST, attributes=("lineno", "col_offset", "end_lineno", "end_col_offset"))
expr = _make("expr", AST, attributes=("lineno", "col_offset", "end_lineno", "end_col_offset"))
expr_context = _make("expr_context", AST)
boolop = _make("boolop", AST)
operator = _make("operator", AST)
unaryop = _make("unaryop", AST)
cmpop = _make("cmpop", AST)
comprehension = _make("comprehension", AST, ("target", "iter", "ifs", "is_async"))
excepthandler = _make("excepthandler", AST, attributes=stmt._attributes)
pattern = _make("pattern", AST, attributes=stmt._attributes)
type_ignore = _make("type_ignore", AST)
type_param = _make("type_param", AST, attributes=stmt._attributes)
arguments = _make("arguments", AST, ("posonlyargs", "args", "vararg", "kwonlyargs", "kw_defaults", "kwarg", "defaults"))
arg = _make("arg", AST, ("arg", "annotation", "type_comment"), stmt._attributes)
keyword = _make("keyword", AST, ("arg", "value"), stmt._attributes)
alias = _make("alias", AST, ("name", "asname"), stmt._attributes)
withitem = _make("withitem", AST, ("context_expr", "optional_vars"))
match_case = _make("match_case", AST, ("pattern", "guard", "body"))


for _name in ("Load", "Store", "Del"):
    globals()[_name] = _make(_name, expr_context)
for _name in ("And", "Or"):
    globals()[_name] = _make(_name, boolop)
for _name in ("Add", "Sub", "Mult", "MatMult", "Div", "Mod", "Pow", "LShift", "RShift", "BitOr", "BitXor", "BitAnd", "FloorDiv"):
    globals()[_name] = _make(_name, operator)
for _name in ("Invert", "Not", "UAdd", "USub"):
    globals()[_name] = _make(_name, unaryop)
for _name in ("Eq", "NotEq", "Lt", "LtE", "Gt", "GtE", "Is", "IsNot", "In", "NotIn"):
    globals()[_name] = _make(_name, cmpop)


_nodes = (
    ("Module", mod, ("body", "type_ignores")), ("Interactive", mod, ("body",)),
    ("Expression", mod, ("body",)), ("FunctionType", mod, ("argtypes", "returns")),
    ("FunctionDef", stmt, ("name", "args", "body", "decorator_list", "returns", "type_comment", "type_params")),
    ("AsyncFunctionDef", stmt, ("name", "args", "body", "decorator_list", "returns", "type_comment", "type_params")),
    ("ClassDef", stmt, ("name", "bases", "keywords", "body", "decorator_list", "type_params")),
    ("Return", stmt, ("value",)), ("Delete", stmt, ("targets",)),
    ("Assign", stmt, ("targets", "value", "type_comment")),
    ("TypeAlias", stmt, ("name", "type_params", "value")),
    ("AugAssign", stmt, ("target", "op", "value")),
    ("AnnAssign", stmt, ("target", "annotation", "value", "simple")),
    ("For", stmt, ("target", "iter", "body", "orelse", "type_comment")),
    ("AsyncFor", stmt, ("target", "iter", "body", "orelse", "type_comment")),
    ("While", stmt, ("test", "body", "orelse")), ("If", stmt, ("test", "body", "orelse")),
    ("With", stmt, ("items", "body", "type_comment")),
    ("AsyncWith", stmt, ("items", "body", "type_comment")),
    ("Match", stmt, ("subject", "cases")), ("Raise", stmt, ("exc", "cause")),
    ("Try", stmt, ("body", "handlers", "orelse", "finalbody")),
    ("TryStar", stmt, ("body", "handlers", "orelse", "finalbody")),
    ("Assert", stmt, ("test", "msg")), ("Import", stmt, ("names",)),
    ("ImportFrom", stmt, ("module", "names", "level")), ("Global", stmt, ("names",)),
    ("Nonlocal", stmt, ("names",)), ("Expr", stmt, ("value",)),
    ("Pass", stmt, ()), ("Break", stmt, ()), ("Continue", stmt, ()),
    ("BoolOp", expr, ("op", "values")), ("NamedExpr", expr, ("target", "value")),
    ("BinOp", expr, ("left", "op", "right")), ("UnaryOp", expr, ("op", "operand")),
    ("Lambda", expr, ("args", "body")), ("IfExp", expr, ("test", "body", "orelse")),
    ("Dict", expr, ("keys", "values")), ("Set", expr, ("elts",)),
    ("ListComp", expr, ("elt", "generators")), ("SetComp", expr, ("elt", "generators")),
    ("DictComp", expr, ("key", "value", "generators")), ("GeneratorExp", expr, ("elt", "generators")),
    ("Await", expr, ("value",)), ("Yield", expr, ("value",)), ("YieldFrom", expr, ("value",)),
    ("Compare", expr, ("left", "ops", "comparators")), ("Call", expr, ("func", "args", "keywords")),
    ("FormattedValue", expr, ("value", "conversion", "format_spec")),
    ("JoinedStr", expr, ("values",)), ("Interpolation", expr, ("value", "str", "conversion", "format_spec")),
    ("TemplateStr", expr, ("values",)), ("Constant", expr, ("value", "kind")),
    ("Attribute", expr, ("value", "attr", "ctx")), ("Subscript", expr, ("value", "slice", "ctx")),
    ("Starred", expr, ("value", "ctx")), ("Name", expr, ("id", "ctx")),
    ("List", expr, ("elts", "ctx")), ("Tuple", expr, ("elts", "ctx")),
    ("Slice", expr, ("lower", "upper", "step")),
    ("ExceptHandler", excepthandler, ("type", "name", "body")),
    ("MatchValue", pattern, ("value",)), ("MatchSingleton", pattern, ("value",)),
    ("MatchSequence", pattern, ("patterns",)), ("MatchMapping", pattern, ("keys", "patterns", "rest")),
    ("MatchClass", pattern, ("cls", "patterns", "kwd_attrs", "kwd_patterns")),
    ("MatchStar", pattern, ("name",)), ("MatchAs", pattern, ("pattern", "name")),
    ("MatchOr", pattern, ("patterns",)), ("TypeIgnore", type_ignore, ("lineno", "tag")),
    ("TypeVar", type_param, ("name", "bound", "default_value")),
    ("ParamSpec", type_param, ("name", "default_value")),
    ("TypeVarTuple", type_param, ("name", "default_value")),
)

for _name, _base, _fields in _nodes:
    globals()[_name] = _make(_name, _base, _fields)

File: crates/piper-syntax/src/ast.rs
//! Abstract syntax tree. Node set and field names follow CPython 3.14's
//! `Python.asdl` one to one so the `ast` module can serve as an oracle.

/// Source span. Lines are 1-based; columns are 0-based UTF-8 byte offsets.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Span {
    pub lineno: u32,
    pub col_offset: u32,
    pub end_lineno: u32,
    pub end_col_offset: u32,
}

impl Span {
    pub fn new(lineno: u32, col_offset: u32, end_lineno: u32, end_col_offset: u32) -> Self {
        Span { lineno, col_offset, end_lineno, end_col_offset }
    }
    /// Span from the start of `self` to the end of `other`.
    pub fn to(self, other: Span) -> Span {
        Span { lineno: self.lineno, col_offset: self.col_offset, end_lineno: other.end_lineno, end_col_offset: other.end_col_offset }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Mod {
    Module { body: Vec<Stmt>, type_ignores: Vec<TypeIgnore> },
    Interactive { body: Vec<Stmt> },
    Expression { body: Box<Expr> },
    FunctionType { argtypes: Vec<Expr>, returns: Box<Expr> },
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeIgnore {
    pub lineno: u32,
    pub tag: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Stmt {
    pub kind: StmtKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind {
    FunctionDef {
        name: String,
        args: Box<Arguments>,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        returns: Option<Box<Expr>>,
        type_comment: Option<String>,
        type_params: Vec<TypeParam>,
    },
    AsyncFunctionDef {
        name: String,
        args: Box<Arguments>,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        returns: Option<Box<Expr>>,
        type_comment: Option<String>,
        type_params: Vec<TypeParam>,
    },
    ClassDef {
        name: String,
        bases: Vec<Expr>,
        keywords: Vec<Keyword>,
        body: Vec<Stmt>,
        decorator_list: Vec<Expr>,
        type_params: Vec<TypeParam>,
    },
    Return { value: Option<Box<Expr>> },
    Delete { targets: Vec<Expr> },
    Assign { targets: Vec<Expr>, value: Box<Expr>, type_comment: Option<String> },
    TypeAlias { name: Box<Expr>, type_params: Vec<TypeParam>, value: Box<Expr> },
    AugAssign { target: Box<Expr>, op: Operator, value: Box<Expr> },
    AnnAssign { target: Box<Expr>, annotation: Box<Expr>, value: Option<Box<Expr>>, simple: bool },
    For { target: Box<Expr>, iter: Box<Expr>, body: Vec<Stmt>, orelse: Vec<Stmt>, type_comment: Option<String> },
    AsyncFor { target: Box<Expr>, iter: Box<Expr>, body: Vec<Stmt>, orelse: Vec<Stmt>, type_comment: Option<String> },
    While { test: Box<Expr>, body: Vec<Stmt>, orelse: Vec<Stmt> },
    If { test: Box<Expr>, body: Vec<Stmt>, orelse: Vec<Stmt> },
    With { items: Vec<WithItem>, body: Vec<Stmt>, type_comment: Option<String> },
    AsyncWith { items: Vec<WithItem>, body: Vec<Stmt>, type_comment: Option<String> },
    Match { subject: Box<Expr>, cases: Vec<MatchCase> },
    Raise { exc: Option<Box<Expr>>, cause: Option<Box<Expr>> },
    Try { body: Vec<Stmt>, handlers: Vec<ExceptHandler>, orelse: Vec<Stmt>, finalbody: Vec<Stmt> },
    TryStar { body: Vec<Stmt>, handlers: Vec<ExceptHandler>, orelse: Vec<Stmt>, finalbody: Vec<Stmt> },
    Assert { test: Box<Expr>, msg: Option<Box<Expr>> },
    Import { names: Vec<Alias> },
    ImportFrom { module: Option<String>, names: Vec<Alias>, level: u32 },
    Global { names: Vec<String> },
    Nonlocal { names: Vec<String> },
    Expr { value: Box<Expr> },
    Pass,
    Break,
    Continue,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind {
    BoolOp { op: BoolOp, values: Vec<Expr> },
    NamedExpr { target: Box<Expr>, value: Box<Expr> },
    BinOp { left: Box<Expr>, op: Operator, right: Box<Expr> },
    UnaryOp { op: UnaryOp, operand: Box<Expr> },
    Lambda { args: Box<Arguments>, body: Box<Expr> },
    IfExp { test: Box<Expr>, body: Box<Expr>, orelse: Box<Expr> },
    Dict { keys: Vec<Option<Expr>>, values: Vec<Expr> },
    Set { elts: Vec<Expr> },
    ListComp { elt: Box<Expr>, generators: Vec<Comprehension> },
    SetComp { elt: Box<Expr>, generators: Vec<Comprehension> },
    DictComp { key: Box<Expr>, value: Box<Expr>, generators: Vec<Comprehension> },
    GeneratorExp { elt: Box<Expr>, generators: Vec<Comprehension> },
    Await { value: Box<Expr> },
    Yield { value: Option<Box<Expr>> },
    YieldFrom { value: Box<Expr> },
    Compare { left: Box<Expr>, ops: Vec<CmpOp>, comparators: Vec<Expr> },
    Call { func: Box<Expr>, args: Vec<Expr>, keywords: Vec<Keyword> },
    FormattedValue { value: Box<Expr>, conversion: i32, format_spec: Option<Box<Expr>> },
    Interpolation { value: Box<Expr>, str: String, conversion: i32, format_spec: Option<Box<Expr>> },
    JoinedStr { values: Vec<Expr> },
    TemplateStr { values: Vec<Expr> },
    Constant { value: Constant, kind: Option<String> },
    Attribute { value: Box<Expr>, attr: String, ctx: ExprContext },
    Subscript { value: Box<Expr>, slice: Box<Expr>, ctx: ExprContext },
    Starred { value: Box<Expr>, ctx: ExprContext },
    Name { id: String, ctx: ExprContext },
    List { elts: Vec<Expr>, ctx: ExprContext },
    Tuple { elts: Vec<Expr>, ctx: ExprContext },
    Slice { lower: Option<Box<Expr>>, upper: Option<Box<Expr>>, step: Option<Box<Expr>> },
}

/// Literal value of a `Constant` node.
#[derive(Debug, Clone, PartialEq)]
pub enum Constant {
    None,
    Bool(bool),
    /// Arbitrary precision integer as a decimal string with optional leading `-`.
    Int(String),
    Float(f64),
    /// Imaginary literal; the value is the imaginary part.
    Complex(f64),
    Str(String),
    Bytes(Vec<u8>),
    Ellipsis,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExprContext { Load, Store, Del }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BoolOp { And, Or }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Operator { Add, Sub, Mult, MatMult, Div, Mod, Pow, LShift, RShift, BitOr, BitXor, BitAnd, FloorDiv }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOp { Invert, Not, UAdd, USub }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CmpOp { Eq, NotEq, Lt, LtE, Gt, GtE, Is, IsNot, In, NotIn }

#[derive(Debug, Clone, PartialEq)]
pub struct Comprehension {
    pub target: Expr,
    pub iter: Expr,
    pub ifs: Vec<Expr>,
    pub is_async: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExceptHandler {
    pub type_: Option<Box<Expr>>,
    pub name: Option<String>,
    pub body: Vec<Stmt>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Arguments {
    pub posonlyargs: Vec<Arg>,
    pub args: Vec<Arg>,
    pub vararg: Option<Box<Arg>>,
    pub kwonlyargs: Vec<Arg>,
    pub kw_defaults: Vec<Option<Expr>>,
    pub kwarg: Option<Box<Arg>>,
    pub defaults: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Arg {
    pub arg: String,
    pub annotation: Option<Box<Expr>>,
    pub type_comment: Option<String>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Keyword {
    pub arg: Option<String>,
    pub value: Expr,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Alias {
    pub name: String,
    pub asname: Option<String>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct WithItem {
    pub context_expr: Expr,
    pub optional_vars: Option<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchCase {
    pub pattern: Pattern,
    pub guard: Option<Box<Expr>>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Pattern {
    pub kind: PatternKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum PatternKind {
    MatchValue { value: Box<Expr> },
    MatchSingleton { value: Constant },
    MatchSequence { patterns: Vec<Pattern> },
    MatchMapping { keys: Vec<Expr>, patterns: Vec<Pattern>, rest: Option<String> },
    MatchClass { cls: Box<Expr>, patterns: Vec<Pattern>, kwd_attrs: Vec<String>, kwd_patterns: Vec<Pattern> },
    MatchStar { name: Option<String> },
    MatchAs { pattern: Option<Box<Pattern>>, name: Option<String> },
    MatchOr { patterns: Vec<Pattern> },
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeParam {
    pub kind: TypeParamKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeParamKind {
    TypeVar { name: String, bound: Option<Box<Expr>>, default_value: Option<Box<Expr>> },
    ParamSpec { name: String, default_value: Option<Box<Expr>> },
    TypeVarTuple { name: String, default_value: Option<Box<Expr>> },
}
